### ISSUE TITLE
CMake suggestions

### ISSUE THREAD

**Issue by nmoinvaz** [zlib-ng/zlib-ng#338]
Had some suggestions while testing integration:

* Turn off `ZLIB_ENABLE_TESTS` by default. I don't think the tests should be turned on by default, since most cases they will only be built when somebody is testing the functionality. With minizip, I have my tests turned off by default. Maybe there is a good reason for it.
* Allow compilation without any of the Gzip functionality. 
  * gzguts.h might need to be moved to ZLIB_GZFILE_SRCS so it doesn't get compiled in with `WITH_GZFILEOP` turned OFF.
  * gzendian.h, why is endian detection part of gz* when it is used in numerous files. Does it make sense to rename this to zendian.h?
* Remove add_library for _zlibstatic_ and don't specify SHARED or STATIC in add_library for _zlib_. Instead, the library type for _zlib_ can be specified by using [-DBUILD_SHARED_LIBS=ON](https://cmake.org/cmake/help/v3.0/variable/BUILD_SHARED_LIBS.html). Maybe this was done to compile both SHARED and STATIC at the same time?

**Comment by mtl1979** [comment]
1. For stable branch, testing should be disabled by default.
2. It should already be possible to disable gzip-related functionality. If that works incorrectly, people need to post error logs as separate issues 
3. Sometimes it is useful to be able to build both static and shared libraries at the same time. Some features require static linkage due to being private symbols. Test suites are one good example.

**Comment by nmoinvaz** [comment]
Other suggestions I have:

* Prefix all the options with something so they don't conflict with other cmake projects. So like `WITH_GZFILEOP` would become `ZNG_WITH_GZFILEOP`.
* On windows, everytime I set up zlib-ng it complains about zconf.h not existing and some other file. I assume this is because the ./configure script doesn't work on windows. It would be great if cmake just detected if it were Windows and copied the zconf.h.in to zconf.h if it were Windows.

**Comment by mtl1979** [comment]
@nmoinvaz extra prefixes is just overkill... We should never depend on other projects.

There is a lot of things that needs to be fixed with cmake but it has been quite low priority as cmake is not something that is native to Windows development environment. Visual Studio 2017 was first to introduce it and it really has issues that make it quite useless and repelling for general use. Personally I use SDK for Windows 7 or 8.1 depending on if I want to make 32-bit or 64-bit binaries.

**Comment by nmoinvaz** [comment]
I'm not saying it would have a dependency on other projects, but that the argument names might conflict with other projects that would want to use zlib-ng. It is a usability thing.

**Comment by mtl1979** [comment]
I'm just saying it is the wrong method for fixing possible issues.

**Comment by nmoinvaz** [comment]
When working in a multi-project cmake environment, it is much easier to tell which variables are associated with which project if they begin with a prefix.

**Comment by mtl1979** [comment]
We try to keep cmake and configure flags as close to each other as possible. It is true there is two distinct differences, but that is intentional. Cmake flags are all upper case and use underscore, and configure flags are all lower case and use hyphen.

**Comment by Dead2** [comment]
I agree that we should add a prefix to our preprocessor defines. Some of them are generic enough that other projects might define them, and that could cause unintended interactions.

A few of your comments are regarding errors that would benefit from compile-logs that show how it happened.

**Comment by mtl1979** [comment]
Preprocessor defines and cmake command-line defines (parameters in other tools) are two completely different things... It would confuse people more if we would add prefix to command-line defines as when they start with "WITH_" it is clear what they do when cmake outputs the configuration results. Extra prefixes would pollute the output unnecessary.

**Comment by nmoinvaz** [comment]
I think it might be best to get rid of the non-cmake makefiles. It adds to the cost to support the project and add new features. Not only do you have to make it for one makefile, you have to modify all the makefiles and test that they all still work on all the various platforms. It is good that zlib-ng got rid of all the old code that zlib had and moved to a modern coding style, but it still is holding on to these ancient makefiles and scripts. It would be better if the project were cleaned up and these old makefiles and configure script were removed. I thought about adding the cmake prefixes, but then I saw the configure script mess and was deterred.

**Comment by mtl1979** [comment]
The problem with cmake is that it is constantly evolving and full Windows support is still unfinished. Things tend to work when using MinGW or Cygwin but fail miserably when using Visual C++ due to not using GNU tools. I use cmake and Visual C++ at work constantly and know what pain in the ass it is even when using latest available cmake version.

**Comment by donny-dont** [comment]
@mtl1979 what specifically is the problem with CMake's windows support? I've built with MSVC on both Ninja and VS without issues. Its certainly useful to have A single build system.

**Comment by nmoinvaz** [comment]
I also use cmake on Windows for many projects both with minizip and various projects at work. Visual Studio support for cmake needs some improvements, however I never use Visual Studio to run cmake. In my opinion, cmake is just there to generate the project and solution files, once that is done, I open them up in Visual Studio.

**Comment by mtl1979** [comment]
@donny-dont On some machines it detects processor incorrectly and that results in wrong binaries being created. There was also some issues that were fixed in last few cmake versions.

**Comment by nmoinvaz** [comment]
@mtl1979 - Are you using the version of cmake that comes with Visual Studio? I've always downloaded cmake from cmake.org.

**Comment by mtl1979** [comment]
When I use cmake with Visual C++, I usually override processor detection on command-line and then just run "cmake --build . --config Release" and that works the best... Trying to run Visual Studio is just waste of time and resources as it will require a lot of memory just to start.

**Comment by mtl1979** [comment]
@nmoinvaz I'm currently using 3.12.0 from cmake download page. I'm still using Visual Studio 2015 and that didn't come with cmake as install option. I can't install Visual Studio 2017 as it has bug that causes run-time errors with some of the programs I need to use and fails to compile some code that works correctly with Visual C++ 2015 Service Pack 3. I know it's a compromise but necessary for me to be able to compile applications on daily basis.

**Comment by mtl1979** [comment]
One of the biggest issues with Visual C++ and cmake is the limited support for cross-compiling and some people forget that the compilers used for 64-bit Intel processors (AMD64/Win64 target) and 32-bit ARM processors are actually cross-compilers as VIsual Studio is 32-bit application suite where limited support for 64-bit applications was added years after the initial release. Like I mentioned above, it is possible to override the processor detection by selecting non-default cmake generator but that is something we should not need or require.

**Comment by nmoinvaz** [comment]
Right now in the repository there are the following makefiles, scripts:

- CMakeLists.txt (cmake)
- configure (unix only bash script)
- win32\Makefile.arm (nmake for ARM)
- win32\Makefile.msc (nmake for x86/x64)
- test\Makefile.in (unix makefile?)
- testCVEinput.sh (unix only bash script)
- arch/arm/Makefile.in (unix makefile ?)
- arch/generic/Makefile.in (unix makefile ?)
- arch/s390/Makefile.in (unix makefile ?)
- arch/x86/Makefile.in (unix makefile ?)

If somebody is adding a feature to zlib-ng they have to:

- Have knowledge of 
  - unix makefiles
  - bash files
  - nmake
  - cmake
- Be able to update all those
- Ensure that they (all, in an ideal world) still work

I've already seen instances where the nmake files are out of date. I argued back then that it would be better at that time to nuke them. Only out of mercy did I test the changes that were made to those nmake files.

I'm not sure I understand the need for all these different things consider that:

- CMake can generate unix makefiles
- CMake can generate nmake makefiles
- CMake can generate Visual Studio and Xcode projects
- CMake has processor and compiler environment feature detection
- nmake is essentially dead in the Windows world (might be my opinion only)

One CMakeList.txt should be able to take the place of all those files. Additionally on Windows, there are issues with zconf.h not being copied correctly and other files missing when using cmake - this is because I assume that the configure script is doing something to create this file, that is not in cmake. So already there is a divergence. 

I understand that there might be a desire to use the "right tool for the right platform" but cmake is a cross-platform tool that is widely used. Please consider what that particular principle is costing in terms of time and complexity.

**Comment by mtl1979** [comment]
This conversation is just pointless... I already pointed out cmake is buggy and does not work good enough to justify nuking better alternatives.

**Comment by nmoinvaz** [comment]
Every large software project has bugs, most of the time they can be overcome in other ways. There are not enough specifics about the bugs you encountered so I can't really speak to your experience - I can only speak to mine. Both in minizip and our complex software project at work, we switched to CMake because the alternative was to manage and maintain project and solution files for the major IDEs across the 3 platforms. It was a lot of work to manage all that - especially when the software was refactored from time to time. We gained many benefits from moving to cmake, even our build scripts became much simpler as there was a single way we built on all platforms - `cmake --build . --config Release`

I don't see unix makefiles, unix bash scripts, and nmake as better alternatives to cmake. Not all of them have as many features as cmake and I think working with some of them is much more difficult and the syntax is a lot more cryptic. The funny thing is that this repository already uses cmake. I thought the project was called zlib-ng, the _ng_ part standing for _next generation_, but seems to me there are still some _first-generation_ legacy makefiles and script that are weighing it down. My suggestion is that we remove all those makefiles I mentioned in my earlier comment in favor of cmake.

**Comment by mtl1979** [comment]
As a major contributor to the project my vote is still against removing Makefiles until cmake fully works with Windows and Visual C++... This also means fixing the configuration detection in CMakeLists.txt ... Just to be clear, we already nuked separate Makefiles for Cygwin and MinGW because we could fix all the issues in configure script.

**Comment by gdevenyi** [comment]
Maybe @hjmjohnson can assist in this, as I would consider him a cmake expert

**Comment by Dead2** [comment]
We can only remove makefiles if there is a fully functional alternative in place, that much is for certain.
First the alternative needs to be in place so it can be thoroughly tested, and after that we can consider the removal of others, we will not be doing the two steps as one operation.

If cmake can be made to be a fully functional replacement, then of course we can consider removing the others. But a bit discussion on removal of the others before there is an alternative in place is premature and is not likely to produce much valuable output.

**Comment by hjmjohnson** [comment]
I am willing to assist, but I have many other high priorities in June.  

I would need a list of tasks that need to be performed.  What is the makefile system supplying that needs to be instrumented in cmake?

- [ ]  also means fixing the configuration detection in CMakeLists.txt  <-- What is broken in the configuration detection?

**Comment by mtl1979** [comment]
@hjmjohnson Basically the current version was designed for gcc, so anything else including Microsoft's cl.exe is only partially supported. Not all can be fixed in CMakeLists.txt as there is known bug in cmake binary itself regarding default generator under VS2013, VS2015 and VS2017. It was scheduled to be fixed for VS2019 when it is released.

**Comment by nmoinvaz** [comment]
@Dead2, I have added the ARM support needed for CMAKE to bring it up to speed with the configure script.

**Comment by mathstuf** [comment]
(CMake developer here)

@mtl1979 Could the issues you've seen with CMake and Visual Studio be filed somewhere? We cannot fix issues we don't know about. If it is specific to the CMake shipped with Visual Studio, there's not too much we can do about that though (I think they're closer to using unpatched versions, but I don't think it is quite complete yet). But Visual Studio support in CMake itself is very well tested, so it would be nice if any problems you're seeing would be reported. Thanks.

**Comment by mtl1979** [comment]
@mathstuf Visual Studio 2015 does not come with bundled cmake, so the issue is with official download. It was already reported as I knew it was gonna be fixed for later version of Visual Studio. The main problem is that Visual Studio defaults to 32-bit Intel executables in all possible scenarios and only way for cmake to work correctly is to honor the environment variables instead of relying to scanning registry contents.

**Comment by mathstuf** [comment]
Starting with 3.14, CMake defaults to using the host platform as the default target for Visual Studio generators. The `-A` flag is the way to switch which architecture is targeted for that ~~platform~~ generator as well. For the Ninja generator, yes, CMake needs to be run in the appropriate environment to get the compiler toolchain you want.

Edit-to-add: Hmm, maybe that's just the 2019 generator that changed that default behavior. I'll ask here.

**Comment by mtl1979** [comment]
@mathstuf Using host platform is incorrect too as it will also be correct in only 33% of the cases. As most of the people using Visual Studio know, it can be used for compiling 32-bit Intel, 64-bit Intel and 32-bit ARM on either 32-bit or 64-bit Windows.

**Comment by mathstuf** [comment]
That is the common case though… There has to be some default target architecture. CMake doesn't really do builds over a cross product of requested flags (ancient internal architecture decisions mean this isn't going to change anytime soon), so each build tree only supports a single architecture. The `-A` flag is the way to select a non-default target architecture.

**Comment by mtl1979** [comment]
@mathstuf That was my point, it's design flaw in the detection of correct generator. It can never guess the correct one in 100% of the cases because Visual Studio is essentially cross-compiler. It can only work if it uses same logic as Visual Studio itself. One batch file setting up the build environment for each host-target pair.

**Comment by mathstuf** [comment]
Cross compiling is not something one can just "assume" in a given build and expect everything to be peachy. Maybe here, but any build which generates a tool that it expects to be able to run later needs to build it for the host. Any `try_run` check needs similar assistance. How are the tests meant to be run?

Other projects have been able to work with the way CMake works today (and I'll be among the first to admit that there are definitely warts). I guess what I'm trying to say is that I'm not understanding what you actually want to happen here short of just shipping hand-done `.vcproj` files in the project.

**Comment by mtl1979** [comment]
@mathstuf Basically we were just comparing which one is better, hand-crafted nmake makefile already in the tree or cmake. Obviously configure script can only be used with Cygwin or MinGW.

**Comment by mathstuf** [comment]
OK, thanks. Yes, if that is what you want, then CMake cannot provide such a build (and is highly unlikely to in the forseeable future). I'll observer that that does not preclude making improvements to the CMake build itself, only in making it the only supported build system for the project.

**Comment by mtl1979** [comment]
@mathstuf I think you understood my point quite well... I don't like bitching about decisions made by other projects but removing something that "just" works and replacing it with something that only works about 33% of the cases is in my humble opinion one very good reason to reject a change.

**Comment by nmoinvaz** [comment]
> Remove add_library for zlibstatic and don't specify SHARED or STATIC in add_library for zlib. Instead, the library type for zlib can be specified by using -DBUILD_SHARED_LIBS=ON. Maybe this was done to compile both SHARED and STATIC at the same time?

Does it make sense to have a `BUILD_SHARED_ONLY` and `BUILD_STATIC_ONLY` cmake options? Or perhaps a `BUILD_SHARED_AND_STATIC` option that is defaulted to ON. When turned off then it will respect the `BUILD_SHARED_LIBS` cmake option..

**Comment by nmoinvaz** [comment]
Honestly, I think the CMakeLists.txt will be better served by just have a single library build and configurable with cmake's `BUILD_SHARED_LIBS` option due to all the branching that would be required - having to do this if zlib building or that if zlibstatic etc.

**Comment by mathstuf** [comment]
The rationale is that on the consuming side, CMake prefers to use the target name. The idea is that the consumer just says `target_link_libraries(mytgt PRIVATE zlib::zlib)` and whether that is static or shared depends on what the found zlib package is built as. Having to do `if`/`else` in the consuming code (possibly based on some `USE_STATIC_ZLIB` option, more likely just short-circuited via `BUILD_SHARED_LIBS`) isn't how CMake projects usually work (remember to multiply the conditional times the number of dependencies I have). Usually you have all-static or all-shared, but that isn't necessarily always the case.
